### ISSUE TITLE
ScErrUnexpMetadata (SCERR7035): Table Palma.ComputedN2Property: kernel reports SCERRCOLUMNNAMENOTUNIQUE

### ISSUE THREAD

**Issue by miyconst** [Starcounter/Starcounter#4092]
The database schema was changed, and two columns from `Palma.ComputedN2Property` class where moved to it's parent `Palma.ComputedNProperty` class.

Starting app produces the following error:

```
System.Exception: ScErrUnexpMetadata (SCERR7035): Internal error in unmanaged meta-data population. Table Palma.ComputedN2Property: kernel reports SCERRCOLUMNNAMENOTUNIQUE.
Version: 2.3.0.5675.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR7035.
   at Starcounter.SqlProcessor.SqlProcessor.MetalayerThrowIfError(UInt32 err) in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Starcounter\SqlProcessor\SqlProcessor.cs:line 379
   at Starcounter.SqlProcessor.SqlProcessor.<>c__DisplayClass57_1.<NewTransaction>b__0() in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Starcounter\SqlProcessor\SqlProcessor.cs:line 453
   at Starcounter.Internal.TransactionManager.Scope(TransactionHandle handle, Action action) in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Starcounter\TransactionManager.cs:line 471
   at Starcounter.Transaction.Scope(Action action) in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Starcounter\Transaction.cs:line 198
   at Starcounter.SqlProcessor.SqlProcessor.NewTransaction(Action a) in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Starcounter\SqlProcessor\SqlProcessor.cs:line 450
   at Starcounter.SqlProcessor.SqlProcessor.AlterTableAddColumns(String full_table_name, STAR_COLUMN_DEFINITION_WITH_NAMES* added_columns, UInt64& new_layout_handle) in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Starcounter\SqlProcessor\SqlProcessor.cs:line 159
   at Starcounter.Binding.TableUpgrade.Eval() in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Starcounter\Binding\TableUpgrade.cs:line 102
   at Starcounter.Hosting.Package.CreateOrUpdateDatabaseTable(TypeDef typeDef) in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 456
   at Starcounter.Hosting.Package.UpdateDatabaseSchemaAndRegisterTypes(String fullAppId, TypeDef[] unregisteredTypeDefs, TypeDef[] allTypeDefs) in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 354
   at Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application, ApplicationDirectory applicationDir) in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 265
   at Starcounter.Hosting.Package.Process() in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 138
   at Starcounter.Hosting.Package.Process(IntPtr hPackage) in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 55
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\TeamCity\BuildAgent\work\sc-2443\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 85
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR7035
```

I know and understand how that happened, but technically this operation is just "add" columns to parent class and "remove" columns from the child class, which we claim to support.

Even if we can't implement support in this situation, it should fail with a proper exception, explaining that it is not supported.

cc @k-rus, @dan31

**Comment by miyconst** [comment]
Related - [ScErrTypeBaseDeviation, removed inheritance from a class.](https://github.com/Starcounter/Starcounter/issues/4083).

**Comment by k-rus** [comment]
> "remove" columns from the child class, which we claim to support.

Actually we don't support removing columns. We support not specifying properties, which leads to the above problem.

I see two solutions for this issue:

1. Drop support for undefined properties.
2. Improve error message that the column is already defined in a table.

**Comment by miyconst** [comment]
>Improve error message that the column is already defined in a table.

This at least has to be implemented.

**Comment by k-rus** [comment]
> Improve error message that the column is already defined in a table.
> This at least has to be implemented.

Thanks for the feedback.

We will look at it.

@alexanderstc 
I would expect to get unhandled exception here, since it is reported as internal metadata. But now we have request to actually fix the error message something like:

> Column MyCol already exists in table MyChildTable and cannot be added to table MyParentTable

**Comment by k-rus** [comment]
I guess the fix should be done for 2.4. Let me know if it should be done for 2.3.

**Comment by miyconst** [comment]
It would be nice to see it in 2.3, since this limitation was introduced in 2.3 and this is what our customers will see quite soon.

**Comment by k-rus** [comment]
@dan31 
2.3 is planned to release on Monday. Is it good idea to do any fixes now?

**Comment by alexanderstc** [comment]
>> Improve error message that the column is already defined in a table.

> This at least has to be implemented.

It's implemented and unittested for star_alter_table_add_columns() long time ago.
Any other metalayer call needs fixing?
Could you provide a reproducible branch?

**Comment by alexanderstc** [comment]
Maybe it's https://git.starcounter.org/BlueStar/level0/issues/60

**Comment by alexanderstc** [comment]
But that one is closed.

**Comment by k-rus** [comment]
> Maybe it's https://git.starcounter.org/BlueStar/level0/issues/60

No, it is different case. Here the column is presented in child layout, not in parent layout.

@dan31 suggests that the fix should go into 2.3.

**Comment by alexanderstc** [comment]
Reproduced.

**Comment by alexanderstc** [comment]
Fixed, waiting for git.starcounter.org to commit.

**Comment by alexanderstc** [comment]
@dan31 fixed my vpn, proceeding with committing and running buildserver tests.

**Comment by alexanderstc** [comment]
Pushed a fix (a proper error message) to v2.3 and develop.
Closing for now, will keep an eye on extended and nightly builds.
